static/js/composebox_typeahead.js

```diff
@@ -349,9 +349,18 @@ exports.content_typeahead_selected = function (item) {
                 + '#**' + item.name + '** ');
         $(document).trigger('streamname_completed.zulip', {stream: item});
     } else if (this.completing === 'syntax') {
-        rest = "\n" + beginning.substring(beginning.length - this.token.length - 4,
-                beginning.length - this.token.length).trim() + rest;
-        beginning = beginning.substring(0, beginning.length - this.token.length) + item + "\n";
+        if (rest === '') {
+            // If cursor is at end of input ("rest" is empty), then
+            // complete the token before the cursor, and add a closing fence
+            // after the cursor
+            rest = "\n" + beginning.substring(beginning.length - this.token.length - 4,
```

Comment:
Thanks for adding the comments.  I still think this code could be made even more clear.  It's strange that we compute `beginning` before `rest`.  And we should extract `beginning.length - this.token.length - 4` to a local var and explain that the 4 refers to the length of "~~~ " or whatever.